src/networking.c

```diff
@@ -1602,29 +1602,93 @@ int _writeToClient(client *c, ssize_t *nwritten) {
+            char *iov_base;
+            size_t iov_len;
+            iovcnt = 0;
+            while ((next = listNext(&iter))) {
+                o = listNodeValue(next);
+                if (o->used == 0) { // empty node, just release it
+                    c->reply_bytes -= o->size;
+                    listDelNode(c->reply, next);
+                    continue;
+                }
 
-        if (objlen == 0) {
-            c->reply_bytes -= o->size;
-            listDelNode(c->reply,listFirst(c->reply));
-            return C_OK;
-        }
+                iov_base = o->buf;
+                iov_len = o->used;
+                /* The first node of reply list might be incomplete from the last call,
+                 * thus it needs to be calibrated to get the actual data address and length. */
+                if (iovcnt == 0) {
+                    iov_base = o->buf + c->sentlen;
+                    iov_len = o->used - c->sentlen;
+                }
+                iov[iovcnt].iov_base = iov_base;
+                iov[iovcnt].iov_len = iov_len;
+                if (++iovcnt == IOV_MAX) break;
+            }
+            if (iovcnt == 0) return C_OK;
+            *nwritten = connWritev(c->conn, iov, iovcnt);
+            if (*nwritten <= 0) return C_ERR;
 
-        *nwritten = connWrite(c->conn, o->buf + c->sentlen, objlen - c->sentlen);
-        if (*nwritten <= 0) return C_ERR;
-        c->sentlen += *nwritten;
+            ssize_t sent_bytes = *nwritten - (ssize_t)iov[0].iov_len;
+            if (sent_bytes >= 0) { // it's safe to release the first node
+                listRewind(c->reply, &iter);
+                next = listNext(&iter);
+                o = listNodeValue(next);
+                c->reply_bytes -= o->size;
+                listDelNode(c->reply, next);
+                c->sentlen = 0;
+            } else { // still stuck in the first node
+                c->sentlen += *nwritten;
+                sent_bytes = 0;
+            }
 
-        /* If we fully sent the object on head go to the next one */
-        if (c->sentlen == objlen) {
-            c->reply_bytes -= o->size;
-            listDelNode(c->reply,listFirst(c->reply));
-            c->sentlen = 0;
-            /* If there are no longer objects in the list, we expect
-             * the count of reply bytes to be exactly zero. */
-            if (listLength(c->reply) == 0)
-                serverAssert(c->reply_bytes == 0);
+            /* Locate the new node which has leftover data and
+             * release all nodes in front of it. */
+            while (sent_bytes) {
+                next = listNext(&iter);
+                o = listNodeValue(next);
+                if (sent_bytes < (ssize_t)o->used) {
+                    c->sentlen = sent_bytes;
+                    break;
+                }
+                sent_bytes -= (ssize_t)o->used;
+                c->reply_bytes -= o->size;
+                listDelNode(c->reply, next);
+            }
```

Comment:
This piece of code can be moved into `if (sent_bytes >= 0)` so that the `sent_bytes = 0;` in the above else can also be removed.

Reply:
done.